### PR TITLE
Small fixes for estimation of christmas models

### PR DESCRIPTION
**Pull request by janosg**
### Current behavior

Too much memory consumption

### Solution / Implementation

- reduce dtype for number of contacts to `uint16`
- reduce dtype of group codes for recurrent contact models to `int32`
- remove the random loop order in the matching algorithm (reduces memory and increases speed)
- trigger garbage collection in a few places (i.e. write `x = func(x)` instead of `y = func(x)` if `x` is not needed after passing it to `func` and readability does not suffer.

### CODE REVIEW

File: src/sid/msm.py [hunk 1]
```diff
@@ -141,7 +140,7 @@ def _msm(
     flat_empirical_moments = _flatten_index(empirical_moments)
     flat_simulated_moments = _flatten_index(simulated_moments)
 
-    moment_errors = flat_empirical_moments - flat_simulated_moments
+    moment_errors = flat_simulated_moments - flat_empirical_moments
```
Comment (tobiasraabe) [inline src/sid/msm.py:143]: Is the order important to manfred? If it's true, I would add the comment.
```suggestion
    # Order is important to manfred.
    moment_errors = flat_simulated_moments - flat_empirical_moments
```

File: src/sid/parse_model.py [hunk 2]
```diff
@@ -56,7 +56,9 @@ def parse_share_known_cases(share_known_cases, duration, burn_in_periods):
             )
         # Extend series with burn-in periods and if shares for burn-in periods do not
         # exist, backfill NaNs.
-        share_known_cases = share_known_cases.reindex(extended_index).backfill()
+        share_known_cases = share_known_cases.reindex(extended_index).fillna(
+            method="bfill"
+        )
```
Comment (tobiasraabe) [inline src/sid/parse_model.py:61]: Can you remember why the old version did not work? They should be the same.

**Review by tobiasraabe — APPROVED** [review]
👍

File: src/sid/msm.py (hunk 1 above)
Comment (janosg) [reply src/sid/msm.py:143]: Yes, the order is important if we want to exploit that infection probabilities have a positive effect on infection numbers.

File: src/sid/parse_model.py (hunk 2 above)
Comment (janosg) [reply src/sid/parse_model.py:61]: Yes. The other version requires a very new version of pandas that is not compatible with estimagic yet.

**Pull request merged**